Changes for v2 paste format

elrido opened the pull request:
Dear @r4sas,

As [discussed in another topic](https://github.com/PrivateBin/PrivateBin/pull/431#issuecomment-497230549), these are the changes to handle v2 paste format of the, as of yet unrealeased, PrivateBin v1.3.

This PR is incomplete in that it doesn't keep backwards compatibility. I will update it with changes that will let your client work with both older and newer PrivateBin installations. I probably will also try and reduce the duplicate logic between versions and en/decryption, but wanted to raise this so we can coordinate this and so that you have an easier way to request changes to fit this projects coding style / proper python syntax / etc. (I am not that experienced with python yet and am happy to learn and improve).

Best,
El RIDO

r4sas commented:
I'll do some checkups and changes in code before starting adding v2 support.

When 1.3 release is planned?

elrido commented:
No fixed time, so no need to hurry. I think we should at least address the >2MB file download issue in Chrome and check if some [other planned changes](https://github.com/PrivateBin/PrivateBin/projects/3) should get included.

elrido commented:
I have merged your changes and introduced a suggestion, how we could handle the version specific logic in format.py. Let me know if you like this concept or if you would want a different structure.

I can provide the v1 logic as well, but if you don't mind, I would replace the use of the python SJCL library in it with straight pycryptodome as I am not too impressed by [that libraries code](https://github.com/berlincode/sjcl/blob/master/sjcl/sjcl.py#L81). There are a few TODOs in important places and adata isn't implemented, but just assumed to be always an empty string. To be fair, that library will work for most pastes ever created with ZeroBin or PrivateBin.

r4sas commented:
I think leave SJCL for v1 paste as is and import it when it needed. But for v2 we sure would create other functions, which will work with PyCrypto directly.

About moving to format.py - we already have api.py, why not use it?

r4sas commented on pbincli/utils.py:
```diff
@@ -23,11 +23,16 @@ def check_writable(f):
+
+
+def json_encode(d):
+    return json.dumps(d, separators=(',',':')).encode()
```
Are we really need such compact json?
Since 3.4 used `(',', ': ')` as default.
https://docs.python.org/3/library/json.html#json.dump

A know, that done same way (by replacing spaces in data), but I dont remember why...

elrido replied:
In most places this isn't strictly required, except for the adata. In the v2 format, settings of the paste that are OK to expose to readers of the paste were moved into the adata field, as that is included in the verification of the cipher message, but stored plain text next to the message. Gives us the benefit that readers can be sure that the message wasn't manipulated on the server.

For this to work, both the creator and the reader of a paste need to generate the same adata JSON string from the paste object. This is why I am using arrays with fixed positions instead of objects in the adata section, since JSON doesn't require parsers to preserve the order of properties, but the order in arrays is fixed.

Except that python by default produces human readable JSON with spaces and you [had previously replaced these with a search-and-replace](https://github.com/r4sas/PBinCLI/blob/1ff6e721c7c562168839d1d029224608a5c4f28e/pbincli/actions.py#L44). When I found the above way, I switched to this as it seemed simpler. In the other uses of `json.dumps` we could skip that bit, but then again I still have to encode the result, we have the library function anyway and even conserve a few bytes sending the data.

r4sas commented on pbincli/api.py:
```diff
@@ -3,25 +3,43 @@
-        r = requests.post(url = self.server, headers = self.headers, proxies = self.proxy, data = request)
-        return r.text
+    def delete(self, request):
+        from pbincli.utils import json_encode
```
Not needed here ;)

elrido replied:
Well spotted, pushed a fix!

r4sas commented on pbincli/actions.py:
```diff
@@ -1,217 +1,150 @@
                 f.close
 
-        if 'burnafterreading' in result['meta'] and result['meta']['burnafterreading']:
+        # burn after reading via API, only required for PrivateBin < 1.2
```
How we can detect if PBin version is < 1.2?

json-ld for < 1.3 is same for any version (< 0.22 I'm not counting).

elrido replied:
That is correct. Hence I propose to send that burnafterreading delete for all v1 pastes just in case it happens to be a 1.2.1 or 1.2.0 instance.

Note for future reference: The version string in the HTML isn't a reliable indicator, as some folks change it in their server and not all templates have it in the same HTML element.

r4sas replied:
I'm asking that because I found that 1.2.1 returns empty response, which causes exception.

And I made exception handler to prevent this:
```diff
-    def delete(self, pasteid, token):
-        request = {'pasteid':pasteid,'deletetoken':token}
-        r = requests.post(url = self.server, headers = self.headers, proxies = self.proxy, data = request)
-        return r.text
+    def delete(self, request):
+        result = requests.post(
+            url = self.server,
+            headers = self.headers,
+            proxies = self.proxy,
+            data = request)
+
+        # using try as workaround for versions < 1.3 due to we cant detect
+        # if server used version 1.2, where auto-deletion is added
+        try:
+            return result.json()
+        except ValueError as e:
+            # unable parse response as json because it can be empty (1.2), so simulate correct answer
+            from json import loads as json_loads
+            return json_loads('{"status":0}')
```

elrido commented:
Reg. format.py - I had understood the api.py to contain all the server interaction logic and this was more about paste format.

Should I move the `Paste*` classes to api.py? Should I provide a V1 implementation in it?

Reg. PyCrypto - Not a Python pro here, but I noticed that it's repo seems not updated since 2014 and the SCJL library used pycryptodome instead. I would therefore suggest to use the latter.

r4sas commented:
> Reg. format.py - I had understood the api.py to contain all the server interaction logic and this was more about paste format.
>
> Should I move the `Paste*` classes to api.py? Should I provide a V1 implementation in it?

I think yes, to prevent increasing count of modules. Let me use code snippets, I think I will do all in one class to prevent separation.
 
> Reg. PyCrypto - Not a Python pro here, but I noticed that it's repo seems not updated since 2014 and the SCJL library used pycryptodome instead. I would therefore suggest to use the latter.

As PyCrypto I mean PyCryptodome, which is used now.

r4sas commented:
@elrido is your test instance accepting pastes created with proposed code?

I'm receiving empty response.
```
Filename:       README.rst
MIME-type:      application/octet-stream
Request:        {'v': 2, 'adata': [['ilmaXeKw0XwaSipcm56AEQ==', 'dFukFdGJ2Qw=', 100000, 256, 128, 'aes', 'gcm', 'zlib'], 'plaintext', 0, 0], 'meta': {'expire': '1day'}, 'ct': 'VBGtOFX9yXJJdDfTqzF9VTmzAjaZHjH2jsB5DwatTNwvo6hxVYuaxSUDMe5m+2Jr2NuR1iLxYudG6mx2WAHdnUM4AXPvD7GwU+O+6skMsrrt1I5nFc8+oIHvlemZK4OutSVv8okQUk9HWv7pmURTXypNSBLTK2D4GMIRcUG7aYzXrsvLeCyHGkTDXEAaTiP/efJ2enChyq0tGezgWE3X0ACIErU/BRgrSQ/IEg4TKezNPj4IhbOnFf/3gvw1L+u39hFiB++oeGvqt1xIZlBFDvyoOlTyGTopX++RSiZYVJnMDhao2iH/Ti4XxHf71IOMdbPGLcIyIMXW1pj/vDjcfJRz2OxkzqHjnJBnfhyj7vIm0bFZnCjq3pyVHo/Nsv8YBUFX318Sf0J+k93rPeGx+HWZoI6HA0+ufqTQLDdK5MuxCl7fy2Rh8BoqhJOxJv1SmD6Kx2wlOKz6oIMNJHNU2wuFYBjeCWq5IIGgcBVq7MREbpO4Q9g9l9b2G5ePa31I4DoKKuzz8sPncrXloAr5qR555hWBoqq1vl3x40VxO8IpPgsSXa//SJAjtO09UiE6yMVJ44DNvlqUp8B4S1xWN4Xyf9O+EhptdPvu5E2fF589QBIkRqxvpMJ5CrgU55kYLhh15Awh5Aslg1UOEsZuwpArA2Y9YChpVBroDwf+GPhzAUqYNaRdYtnR4Equ+xL8DrXd7lRF3rBkdNkbsHY41L+Vgo3yPf/odQWXXi+9EECijQVpJSmu3/qB6yDtSVFmT1dgUUs0HS6gRac4UFSv+XXV/gwkji9NpZO2g20YFhOwlB4sCgazTXPy93zwJ0zzi/JBSdw+P1Ph3zRjT7w8yp95jErJSSVL9GnhTiFurJFmZPvngSnzgHagj+j10cPGByCdjkPw6t68jGfrMMtJSjyCxR/IOGRBpU6/whjNRldktkotS0FBD+7a0h2eHOIsiIC07dfakP8fX0pwEaRh1k9nvtxCVwsYE7l5lNylZfTONmByt88zbVtPm4pru41JIMaa7kx5K3UHvxLMhB/P1C8e+H41CGYSRn74bkmx3MFbRmfHg/Pvg8CqhPN4Tli4s9qhXTi7uXsmnnAzDN+sBtrncw1hxVQ7WWcGR/s6z4f6xcUaZdEtKslVpqFpYDRCFvuxkP1xuwzI8XYla8UNfn+3YXWtHyewbMjtEMroNDQ3HatQSbA3R9A1DS/7J/i4RZRUyK0ZfN2u8jJvR8mqYY1gLu8M+qCK7F0+nIdCUZR6i+g8tO8jGjXqqga/IxRXOZ05WCYHz/zHs9SXc5TsEh0DArT/feThcH8fGGp/jKVKQW6g1R1Pn7K9O3mrBZyYkWxNW6z4lM/UekZlc+dy5gGcmNvCaHNy0fzlGtXiHlKueqgP5Sd9d5Li1HcKGys98PaXF5X+etSJgxtoJ3b+igeaLyq8O3WHUQlnGV2YlJ7J8S+QvHHG1xuypBs+8e8uS6+AuOe0GDGVRCJHjtSeTA6F0graFnm363pd4B/1BeCxoFrISiRuCZowDmav/8hyENUcXArDkVbykSzxJeKs73NlhbE/mCHrV+Qf0kkYFJ7eeuCGpgQZO0pDzJa7Dmf4ZMeGwbeEfTBPYlX4eblwktmmku1fZMfvR39Q512vfABgBEM30pdDTzOMX/juHvwAupyhZv0n474UJqgFmVSH7gqUwUNnZaj+XygpK5lD0Fok7UIe40hDWHx5lZNmfCXL4eJLaEbKcQP6cJBIeI+kYXcsMkkdJBe6yCmnaXRjnfAA+3xILltjUYzFD/xg7gmx42Uu+NuiDpN+fLeqe7hdreqX65y9qPZ3X4aEVA9QVHUEk/zMBGbx0nMJw8KdbKqkEv3Y1VLJQN2cI+cXPGN9TxfZHCd0bROvrPDnxxWmco7a0jWRTLz198QayfCSpRFNLhFy+GyqGB9z8SIq4tbF7x6NYfTl5v0ifuC564qRLo+lk9RwBOEIT7+d6IY6J3vq0zippawumduSxPf19xqRLwL2DASUfVZUmvi257BbVOSfYsQeOqr1rjW2N/NIaY6nx6AWfqkipLGPmkjApKtu86mSMskEzy0tYfaXuA6ra8SsTkbGLBkVzdPConB2MW17VGec30y7H5qCaW8LoEr26NiWm81ep6fRO784uh81b5yBRjpI84vtlyJ66YDDBjbPfR4lWs0PU0YD8+G9alSZigvvD34bCItwTMSfcA3zlc2obLIf80bouDIjVihXP8dyV0YjTAO5mBJ67NYuU4Plk+yuJeQQ8A3MdidiJezX5L99oltfkaXTrVCHBHF94bD1rYXRGlsQZh3cygie7fu2UKu2UICL+vFmp4oholUTj0Fr98+vKDP/uGnMr66Fo2PByPa29xB/ErccYKqLvUcc5kTfHlqmGxup/M3EARxY/066Lx7+6C5uM3RqY9Utb3z8qGGK65VJj2iCdVCi6PXraJp+iveqvMOMH76eE1A7jWsc/4NSp9Xq9pe1syHvHr5IUUn4dSGjZ/JxIKwcNomFLjZqAwHWOeTsJKkrOkXiZNUfyLg2BiaJxdkCQqnnOpR2PvrFFOPFNsHhGB/SocfDFK64N4ya2NctcxQPaNCcTEg/DOscsnwM7TO4tfyEcBpQ8MVowXeMqvg/aUBhXl0fh6Os+++ZaSjkvOee0yTBXafm6oi2z7QCGdMGy6iolb1WGoYnbFAg1bGQgiJG9zvRagnGZL6lViIBEfwJ6WydPoSuS9xTceBmuh7j2cwIZPAxLwDCoZ4eV3iXbOP/makQyfrIQeK76c4NZksllFLd6TO2orRAiaV7gWIRO5m5mT1FOi43YdyJRRW9g0MYVSAtgFr4O0LOd6YDXOOhNNLkdtBJocQxvy6mVt0P+vQz+fCXkES8uok9vO1yiyizLZwAx4BZV2iadSVxU49st+na+9mXgoOqWhz85nee8jeDjcNt+siRs9fIUCQONqsEoejAn6hgHRwo5w6SGILcJ+dJc70SiO7Ruu+jp2wH1WNQhXTzU9MHt0VRgb2jyiyx1H/nCONIXnhs5r1U855qGg20fa9xFbqID3BpkhcmyKnkttJbPr4nOu4PJZyIbgNhQy9XwEMrZ5ifTHu8NYqSjW52im/dZA4fRPRUk6+/VKOOSlvnc0OkcITQI2lVqaGFKmp/EUOcPN5xeIqV8lU77gcgN3v6rZamvGWTwnsV2k+0+a/Ng8N2VKiIVVsMf5Pdf7R+tnEr4wjo6MqdChCpXLUuxaYibWd4TMYfc7hncWmeX1F57CFTqYfd2uWWbofqNSX5Yn5FpnN0YjFjZf5RHgqlNNzHkhrWnd92N+wh6fGF9sFoMJaRKk8zA7DSeXy9FTYoPEbPutGIVdWFI7eOeRcI7t+R4bRu9W3KoYDtJwp1+rRkWB5K66j1NPJZdglPWTnzuj5oZtu4TH1z1Lvfve20aYLiJGvUOiIbqKqysB4EIyKL/QlStjXljhaZQpWJ7hfUQ8xF0FS03FUpAodX8EgRkI20cwHtS3j8qU6/UHmHnY1TTPVzQe8MoZ5ZFc/rqhtIsD9QBOVHrAFNViZIOO0dbwE3yn85jBLxzwWhUz2DwcVgz7mEjVHbOtWgaQECW0fqxlLWM0oSteSrSYdFFQtAyVRXVwB5NlUqJVUaYRn0DR3Tc5iC0cO4V9pOcRVkS1uypC8ErJ7Y73tlG/aPdRX1AAe/xWxt5LZmbd1a7mU2gn6Oo5H3RzHxm2klD8pj0icWhtLuSq1HU/gl4EohxMteTufCzBJSXMFtRdCbA6G4RtyquDXUrHjJAwjVlaoNHUAJnwXQeE4BHajxNOnN7oQ/dTYD7Cj2Mp1oZdwaogrve4s59kp176NeRBbBqv07/45DKx7YWXK9lxAxTibJbBQ5RHo49q7w8zyH/sfhsxAJWx8wrkgWpJXZEEHk5yTbF4LoEt1xZ2Q2N82JBPkg48AgUOxMlaeNgjFIPOOhmQU='}
ERROR: Unable parse response as json. Received:

```

But currently I'm compressing data as earlier with rawdeflate.

upd: same with v2 zlib compression proposed by you.
I can push code which must work with both versions, but I got issue with uploading. Currently tested only with 1.2.1.

elrido commented:
I had tested my changes against the test server, but maybe I used the module incorrectly?

Here is an example URL created using the changes in this PR, containing some text and a file:
https://zerobin-test.dssr.ch/?73ef4917256b0d0f#135nFoBkdmnMqKverskUJrL6rgShAJqBh9f1z5pKYV1g

I created it using this command (in a venv, after installing the requirements into it):
```shell
$ echo 'Hello World!
> 
> This is from python.' | python -m pbincli.cli send -f /home/elrido/Bilder/Minutensprunguhr_animiert.gif
Paste uploaded!
PasteID:	73ef4917256b0d0f
Password:	135nFoBkdmnMqKverskUJrL6rgShAJqBh9f1z5pKYV1g
Delete token:	e8c60a2d9b312b5e53d50493f6b11e5360bf353d738df078d21cbde7fef071be

Link:		https://zerobin-test.dssr.ch/?73ef4917256b0d0f#135nFoBkdmnMqKverskUJrL6rgShAJqBh9f1z5pKYV1g
```

r4sas commented:
Okay, let me upload my current code, maybe you will find where I mistaken...

I will upload `Paste` class in same as proposed module name `format`.

r4sas commented:
Decryption now works without problems.

But why server send empty response?

For check and decoding:
```
Filename:       README.rst
MIME-type:      application/octet-stream
Passphrase:     E6JPuDmiKB5PZC8tpdMsEQayxWPzhBfbGjngfnmPd1rr
Request:        {'v': 2, 'meta': {'expire': '1day'}, 'ct': 'GZr8/PJB3zDTZKEWJzPK8fog9DjP+QiOFl8fHAcAhMnpo1ZJy2VT6oOcaMTa2Xa9fFA6pWbk8mhRStQQ5f4saK6e419Ol8X2WKkv4qtQCk0E4RonY39R84rCZrishvNddIFbKidWmJ+cqRmhUCud3tNtsMZxUbu6eHsogzInau10WqX+GDEWLvEDz/ljn2q1hbz5cJDwipvk2tgddQJ7/YwUJRrK0yWNl+v0WdW37qZ76aV4PaCoDwp/qXtBF8Uo89uE1JWajdco6x2Oy5vH+HA6vab3G4lvFFKEtkiOuMu04FWcdX4Co/ymGdqIYRx+V7NQGyu2A6Qx1MaHXbVxFd6IOqRHJDicZczHqta3L5UEya7+GY5z/t+2urmrxyvzxN3I/SJCNV06pJmfMjl9/eMYOGTvVTlsNfxkbpoJT2BVsolZDyFiMgSKlDbEzNCSG65QsHk7syfa6e4EX0sh/VP/zrTLLlCSDv8fvI10sdw8l4X1/dtL9RynglbTb3rkKCg4BJ9AtHy/OLzVhEe8y1VNogA3qoXC9nqXnjOAeFoB+uY9rV5gR4OHqhhq4o6b6WYV8CejdsJTAzYwxfm3W9h65B/26JT2SOajtFxVKsxXZbTrigMmABnsI7Rw8lGGRBXktd4UpGzmaF1N7E/Bvk+D6UAl9oz5+wPzw40Hksg08hdlFXpcTthQpCMROwGDpoWUjy+mpHpUBbRQ75BQZc4AmjRqoSJ4giClYIgxq6MkLh8YTmb1M/L/4iVJ4j6kcPadvARJ/L8z7pT5zF+BV5/FlaBwSH7t4ASD2BHI9MNhOXH46xnAk4CtQKU0vLlq2D55Ol6kQQpE34vsH1Eq3m8GyYYUdG1WuA4nV6xcbhsFz5ZcIuM7geQf1UNslXjGW7YtghjJ0ntBnVVGlnquvs5Gae2Z4/+J/k4h/Tvf5gUwrkL19A63k51Og7O9KrqYHfxa+vh2Rnpwxpb5C5qGu9dwIBqvkRD1OGMAjhFNp7o+fPYj7oXWP2u3AV/kWfxUER/vF8aJKhgcQsbXVKx0lwn3o8F0BUrKHi+y/r8YqDutFd0YXoA5uaND95oMNhoqSp/PzOqAFC6Usf8Eg9CewrmpsvDSWEAQzNupB1hEEHdBhLLoTNghNdofFIZdqc+4XUW4BjlhqZBy7rEluV6eJNzxOGm+zWuMd7PISPheAKNUI4EGdZyXIwHXaH+RTiONoa6rIwnopuTs1swhPWUG8kGsMxIUpQJvwivFsRbvo0vr9XXPpcOCSojsb6cfpQwcuRZNIzPW6BZUQ4ZsEwK+LXM87BnH5A7UteWhlaBTcU6ApMN1T5NdVRJkT/la7iVCFDoLRt8kvlu6pMVdGyNDLZs61bh1PmhqiSkQojId3dbkDfsGNfyFYOlo91OaSZWQDs+XW/mne/EOhNPd1a+Le73ZSPxy03DoJ+QJKh0Z8nY08FaZSXcNj2XEabaIOEn86YSYjjJz/BIG14dHWvsO5xoOiJtRAMrg+Zdlp4oZwEWLiDeKznUK6XaJ7bZ++YOdzMYGNEwkSQcYgVuhIA+kUfJ0f+LjemEXmeDu5Wbo9fMvwDWw8bx5cvURtM8EoNJGsLxMcHv6+VQG2ft6vLMYFNWSLYkQ/HyHKUI0skVfd9BkCVe0A7uqkd6LfXQ6+EKWWTK04mvO1J9giU8cogZPSmoBLDSrdqir2whD6BoQWQ6DsDGjnKBp0o45EPOX5w3VrxtMoC+I2fVrpTpkFa4cFt+C+G0/o7cMBa8atYXvj79YzQzvTVOhorg2k3hhZf61Vnk2IPN5OtUZeLruAaX5aTwQ/PgnO94+DKGKkWaQ0GGilE3lLotSvRgdzyxmwRReuiRWM7EorFMndhLzOc52IT+GsZqFdO7Wcp+dC5F91HISmVEmtd/bpGUOEc+inQnMFDFoms9oG3LV8363MrICE19o6YyGzkotLni8FoXM6clWY/Y/unABQ/sgT5sqyxd3XGHR2hm/1xCEdw==', 'adata': [['IN9rlDTudoRdXR30UOpiGQ==', 'hWAHgGm7rt4=', 100000, 256, 128, 'aes', 'gcm', 'zlib'], 'plaintext', 0, 0]}
ERROR: Unable parse response as json. Received (size = 0):
```


Checking with that code returns correctly decoded data:
```python
>>> data = {jsondata} // as printed in response
>>> passphrase = "E6JPuDmiKB5PZC8tpdMsEQayxWPzhBfbGjngfnmPd1rr"
>>> paste = Paste()
>>> paste.setHash(passphrase)
>>> paste.setVersion(2)
>>> paste.loadJSON(data)
>>> paste.decrypt()
>>> paste.getText()
b''
>>> paste.getAttachment()
<attachment>
```

r4sas commented:
Found why it not worked, forgot encode json which returned by getJSON :laughing:
Took encoding from proposed by you code.

r4sas commented:
Closing PR now, thank you for your code, which was used to add v2 support.
I will tag 0.2 soon and publish package in PyPi (currently it published in TestPyPi).

elrido commented:
FYI and future reference: I moved my original commits sideways into a branch [`v2-paste-format`](https://github.com/PrivateBin/PrivateBin-Python-Cli/tree/v2-paste-format) and replaced the master in the fork with yours. Will now make that fork read-only and link to this project on it.
